### ISSUE TITLE
Feature request: rulesdir option in .eslintrc

### ISSUE THREAD

**Issue by silverwind** [eslint/eslint#2715]
I'm using [sublime-eslint](https://github.com/roadhump/SublimeLinter-eslint) as well as the CLI itself where I use `--rulesdir`. sublime-eslint does allow CLI switches, but only on a global basis, not per-project (which I think is more of an SublimeLinter issue). This presents the problem that sublime-eslint would fail on line 1 of every file because of the custom rule being present in `.eslintrc`.

Ideally, I think every CLI options should have a counterpart in `.eslintrc` to enable full portability, with CLI options taking precedence if both are present.

Another alternative I could see is to not hard-fail when a custom rule isn't found, and continue with the other rules present, but it surely isn't optimal.

**Comment by ilyavolodin** [comment]
We are planning to discontinue support for `rulesdir` in version 2.0.0 in favor of plugins. As such, I don't  see us making any changes to the way that flag works. Sorry. If you want to use local custom rules, I suggest you look into `npm link`. It's much more robust solution.

**Comment by silverwind** [comment]
Looking though plugin docs, I wonder if it's possible to link a plugin directly into `.eslintrc`, so tests can run offline and don't depend on npm (which may not be working/avaiable when we run eslint).

**Comment by silverwind** [comment]
Hmm plugins *could* work. What does 'Before using the plugin you have to install it using npm.' really mean? Can I specify a standalone js without any ill effects?

**Comment by gcochard** [comment]
It just needs to be in the `node_modules` directory. With npm at the latest version, you can specify relative directories in your `package.json` file, so you can install from abother directory in your own work tree.

**Comment by silverwind** [comment]
But what if there is no `package.json` or `node_modules` (this is for io.js)? I'm thinking this whole plugin thing isn't really right for standalone project-specific rules that you don't necessarily want to publish to npm as they have no value to other projects.

I hope you're not going ahead with that `rulesdir` deprecation.

**Comment by silverwind** [comment]
@ilyavolodin care to point me to the reasoning for the `rulesdir` removal? Why would you remove such a useful feature and force people to publish their (possibly private) project-specific rules to npm? It just doesn't seem right to me.

**Comment by ilyavolodin** [comment]
Ughh.. It's been discussed a while back, and I'm not sure I can find that issue now. Basically, `rulesdir` was created early on, before we had support for plugins ready. Now plugins can do the same thing, and npm is a standard way in node to handle dependencies. So having two different ways to accomplish the same task doesn't seem to be a good idea.
To answer your previous questions, yes, you do need package.json. Once you have that, you do not need to publish your rules to NPM, you can use `npm link` command to locally link a directory in your project that contains package.json (this is also a pretty standard thing to do). I don't think that having an extra file is huge enough overhead to warrant us keeping `rulesdir` around.

**Comment by silverwind** [comment]
That seems to be a bit of a narrow view to me. Not every project is a package (like io.js) and not everyone will be comfortable publishing their rules to npm (yes, there are private packages now, but those aren't free). As I see it, plugins are suitable for *multiple* projects while custom rules are for rules specific to a single project.

In my opinion, all options defined on the CLI should also be available in `.eslintrc` (and vice versa) so the options can be truly portable.

**Comment by ilyavolodin** [comment]
I do not think you read my reply carefully enough. As I said, plugins do not need to be published to NPM. All you have to do is create a separate directory within your project for your custom rules (something you already have to do with `rulesdir`). Create a file that will contain your custom rules, and create a package.json file in that directory. After that you can use `npm link` to let npm know that when you are talking about specific package, it should look in this directory instead of fetching something from the internet. This will work just fine, even if you are offline.
So in nutshell, this is exactly the same behavior that you would get with `rulesdir`, but it gives you more capabilities, like being able to create not just custom rules, but also custom processors, and setup defaults for your rules. From my point of view, there are no advantages to keeping `rulesdir` around, other then you need to create one less file.

**Comment by silverwind** [comment]
Indeed I misunderstood you there, sorry about that.

When I read about `npm link`, the first thing that came to mind is that this won't work on Windows (symlinks require administrative priviledges), and I don't think we're going to add a `package.json`.

Another idea, what about allowing paths in the `plugins` array? Like

```
    "plugins": [
        "eslint-plugins/plugin1",
        "eslint-plugins/plugin2/plugin2.js",
        "plugin3"
    ]
```

plugin1 could be a regular npm package, plugin2 a custom unpackaged plugin, and plugin3 could be an public one.

**Comment by gcochard** [comment]
@silverwind @ilyavolodin You don't have to use `npm link`. You can use a [local dependency](https://docs.npmjs.com/files/package.json#local-paths), and `npm install` will put it into `node_modules` as it should.

**Comment by ilyavolodin** [comment]
@gcochard Nice, I haven't use that feature yet. I only used local path to .tar files. So that makes it even easier to use.
@silverwind I used `npm link` couple of times on Windows machines, and it worked fine. Granted, I do have admin access, so that might be why. But as @gcochard noted, that might not even be necessary anymore. I think your proposal is outside of the areas of responsibilities for ESLint. I don't think we should try to replicate functionality of dependency injection libraries inside a linter. It just doesn't feel right.

**Comment by xjamundx** [comment]
Yeah I'm with @silverwind on this. 2 reasons:

1. Projects like io.js literally cannot rely on having any `node_modules` folder
2. It's much easier to teach new users to ESLint how to use rulesdir.

Not sure if this discussion is done done, but I think there are some valid reasons to get this feature.

**Comment by nzakas** [comment]
The discussion is done. There are a lot of problems with `--rulesdir`, not the least of which being naming collisions between custom and native rules. It was a hack until we figured out something better.

If someone wants to propose a new way to load custom rules in a config file that avoids this problem, feel free. In the meantime, `--rulesdir` isn't going to be removed until 2.0.0, and there are more important things to worry about to get 1.0.0 out the door.

**Comment by silverwind** [comment]
I think plugin loading should follow `require` conventions. Either specify by name (which loads from node_modules) or by relative path (starting with `./`). I can look into doing a PR if you'd agree to this.

**Comment by nzakas** [comment]
Before you write any code, please file an issue about extending plugin loading so we can discuss the various details.

**Comment by KnutHelland** [comment]
Just wanted to give my support to silverwinds suggestion here. It makes a lot more sense if plugins follows `require` conventions. It's a shame that it is impossible to have eslint-rules inside the actual project where they're used. It is very useful to make some very project-specific rules, and it is really annoying that it needs to be put into node_modules which is usually ignored by git.

**Comment by cletusw** [comment]
I tried the plugin approach but wasn't satisfied with the fact that other devs had to `npm install` to get updates to rules that were already up-to-date in their local repo, so I've created a plugin to workaround this issue: https://github.com/cletusw/eslint-plugin-local-rules

It's hardcoded to look for a file called `eslint-local-rules.js` at the root of your project because I couldn't figure out a good way to parameterize it. PRs welcome! :-)

**Comment by dtinth** [comment]
We at @taskworld also need to use project-specific rules.

Putting it as ESLint plugin requires reinstalling each time we update the rules, we prefer to have it it the codebase, so that these linting rules can be updated in lock-step.

We ended up monkey-patching Node.js’s module resolution system to rewrite `require('eslint-plugin-proprietary')` into `require('./lib/eslint/plugin')`, so that we can treat our ESLint rules as part of our codebase.

Here’s how we did it:

```js
// .eslintrc.js
const Module = require('module')

// Hack Node module system to recognize our plugin.
Module._resolveFilename = (original => function (request, requester) {
  if (request === 'eslint-plugin-proprietary') {
    return require.resolve('./lib/eslint/plugin')
  } else {
    return original.apply(this, arguments)
  }
})(Module._resolveFilename)
```

This ~~solution~~ hack meets these requirements in keeping our codebase cohesive:

- The ESLint rules shares dependencies with the project (no separate `npm install` step for rules).

- Developer can `git clone` the project and run `npm install`, and the rules will just work, with no further configuration.

  - Without having to commit any files into the `node_modules` folder.

  - Without having to run `npm link ./lib/eslint/plugin` manually.

- Rules can be updated instantly (e.g. when running `git pull`) without having to run `npm install` again.

- Works with CLI and text editors (tested with Atom and VS Code) without requiring any special configuration. (Not possible with `--rulesdir`.)

## Proposed solution

Above monkey patch is a very dirty hack in my opinion, but it worked very well for us and we’ve been doing this for half a year. Therefore, I propose allowing the user to specify a plugin by path like this: e.g.

```yaml
plugins:
  - react
  - proprietary:./lib/eslint/plugin
```

Here, ESLint would load `eslint-plugin-proprietary` but from `./lib/eslint/plugin` instead of `./node_modules/eslint-plugin-proprietary`. Then the rules can be added under `proprietary` namespace.

```yaml
rules:
  proprietary/thou-shalt-use-bem-classnames: error
  proprietary/thou-shalt-use-sanctified-date-formatting-methods: error
  proprietary/thou-shalt-not-use-expressions-in-i18n-functions: error
  proprietary/thou-shalt-not-use-deprecated-internal-apis: error
  proprietary/thou-shalt-not-use-icon-classnames-directly: error
```

If many people like this proposal, we can create an issue about extending the plugin system as https://github.com/eslint/eslint/issues/2715#issuecomment-110438070 suggested.

**Comment by zertosh** [comment]
I would like to revive this issue in light of "Breaking: New Linter API" https://github.com/eslint/eslint/pull/8465 and the current state of loading plugins by path (https://github.com/eslint/eslint/issues/3458).

`--rulesdir` is a great approach for project specific rules. It's used by ESLint ([link](https://github.com/eslint/eslint/tree/a8e1c1ccf2a212612f8aa63d7463dfa9acd920b6/lib/internal-rules)), Node ([link](https://github.com/nodejs/node/tree/917f86ea353e087af3bc4553219ffabaa2193d20/tools/eslint-rules)), and extensively at Facebook. Unfortunately, `--rulesdir` is only available as a flag, and not something that can be specified in an `.eslintrc`. This means that running a plain `eslint` command doesn't fully work. So IDEs and tooling don't work without additional per-project configuration on the part of the user.

The way we've dealt with this at Facebook is by using a js configuration that uses private APIs to simulate what `--rulesdir` does:

```js
// .eslintrc.js
const Rules = require('eslint/lib/rules');
Rules.load(path.join(__dirname, 'resources/eslint-rules'));
```

In ESLint v4, this won't work anymore due to https://github.com/eslint/eslint/pull/8465. 

The proposed alternative to `--rulesdir` is plugins. But that has its own set of (harder) problems. For one, they must be in `node_modules` - `npm link` and `file://` create other problems, if anyone is interested, I'll delve into them. The idea of loading plugins by path would solve the `--rulesdir` use case, but that discussion is stalled. I don't think that problem is solvable given the requirement that it addresses name collisions.

This discussion was ended with:

> The discussion is done. There are a lot of problems with --rulesdir, not the least of which being naming collisions between custom and native rules.

I would argue that those problems are no different from say having a typo in your `eslintrc`. It's part of your own custom setup, so it's on you to fix it. I'd even go as far as to say that being able to override a native rule is a feature, and not a bug.

I'm really concerned about what I'm going to do roll out ESLint v4 internally at Facebook, if we can't make plain `eslint` commands just work on projects with custom rules. I'd really like a `rulesDir` config that has the same status as `root` (only available in local configs). If this is something that would be acceptable, I can submit PR.

**Comment by not-an-aardvark** [comment]
I think the original motivation for not having allowing `rulesDir` in a config file was that we want we don't want there to be a difference between "top-level" configs and shareable configs. The idea is that users should ideally be able to move their config file to a shareable config and have everything still work without modifying it. The `root` option is a bit different because it's a directive for how eslint should look for a config, not something that affects the semantics of the config file itself (so its behavior in a shareable config isn't really relevant).

Assuming we keep that goal, we would have to make the `rulesDir` option work in shareable configs, which seems like it could end up being confusing. For example, two different shareable configs could end up defining the same rule with `rulesDir`, and without a plugin prefix one of them would end up silently overriding the other.

The other option would be to only allow `rulesDir` to work in top-level configs. This would probably work, but it would also make it complicated for users to convert their project's configs into shareable configs, because at the moment this requires no changes to the config at all.

I'd like to find a solution here that works for everyone -- I've also run into this problem myself when dealing with editor integrations. The underlying problem is that we don't have any conventional place to store run-specific config data where other tools can use it. This also prevents us from doing things like putting a `fix` option in a config.

---

Maybe the best solution would be to have a separate of the config file that holds run-specific options, to avoid confusion about which parts of a config can be inherited:

```yml
rules:
  # ...

topLevelOptions: # probably needs a better name
  rulesDir: foo/bar
  
  # ...maybe other top-level-only options
  
```

Everything in `topLevelOptions` would be ignored when the config file is not the top-level config file for a project.

**Comment by not-an-aardvark** [comment]
By the way, as a workaround it is still possible to achieve something this in ESLint 4 with monkeypatching (although the monkeypatch changed a bit from ESLint 3):

```js
const Plugins = require('eslint/lib/config/plugins');
const loadAll = Plugins.prototype.loadAll;
Plugins.prototype.loadAll = function(...args) {
  this.define('your-plugin-name', require('.'));
  loadAll.apply(this, args);
};
```

**Comment by zertosh** [comment]
> Assuming we keep that goal, we would have to make the rulesDir option work in shareable configs, which seems like it could end up being confusing.

Agreed!

> Everything in topLevelOptions would be ignored when the config file is not the top-level config file for a project.

This is perfect. I've always wanted to configure `max-warnings` and `ext`. And I know some people that would love to set `no-ignore`.

@not-an-aardvark would you accept a PR for this?

**Comment by not-an-aardvark** [comment]
As an addition to core, this would need to be accepted by the TSC (see [here](http://eslint.org/docs/maintainer-guide/issues.html#evaluating-core-features-and-enhancements-tsc-members-only)).

cc @eslint/eslint-tsc Any feedback on this? Hopefully we can address any concerns and then vote on the proposal in a meeting.

**Comment by gyandeeps** [comment]
I want to understand more like how people are using ESLint where they are facing issue because  according to our standard API using [`CLIEngine`](http://eslint.org/docs/developer-guide/nodejs-api#cliengine) you can specify `rulePaths`. Through [`CLI`](http://eslint.org/docs/user-guide/command-line-interface#--rulesdir) you can use `--rulesdir` option. What am I missing? 

I know in the past, we have talked about deprecating `--rules-dir` option in favor of plugins.

**Comment by dtinth** [comment]
@gyandeeps Text editor integrations.

p.s. I think this issue isn’t about `--rulesdir` option anymore. Rather it’s about how to put in project-specific rules as part of the codebase (without having to create a separate plugin package).

**Comment by zertosh** [comment]
ESLint 4.0.0 came out, and never got an answer on this from the TSC :(

**Comment by gyandeeps** [comment]
I am still confused a little bit... can you please provide answers to https://github.com/eslint/eslint/issues/2715#issuecomment-306079492 question so that I can better understand the problem here. Make sure i am not missing anything. thanks

**Comment by silverwind** [comment]
> I am still confused a little bit

My use case would be editor integration. Editor plugins are usually dumb and just call something similar to `cd project; eslint file.js`.

Now projects like Node.js maintain a set of additional rules in a [directory](https://github.com/nodejs/node/tree/master/tools/eslint-rules) in the repository, so eslint must somehow know about these rules, and I thought the configuration file would be the best (and likely only) place where this can be specified.

And yes, this isn't only about `--rulesdir`, I could see `--plugin`, `--cache-location`, `--ignore-path`, `--no-ignore` and possible more options to benefit from `.eslintrc` equivalent.

**Comment by not-an-aardvark** [comment]
**TSC Summary**: This proposes to add a `topLevelOptions` property to a config file, only available in top-level config files and not in shareable/extended configs, to allow top-level options such as plugins and extensions to be statically analyzable by tools like editor integrations.
**TSC Question**: Should we accept this proposal?

**Comment by not-an-aardvark** [comment]
There weren't enough people in the TSC meeting today to reach quorum, but there was a brief discussion about the proposal ([archive](https://gitter.im/eslint/tsc-meetings/archives/2017/06/15)).

The tl;dr is that we should probably create a new issue for this proposal, because there have been several different proposals in this issue and it's confusing for people to read it and figure out what's going on.

**Comment by zertosh** [comment]
> The tl;dr is that we should probably create a new issue for this proposal, because

@not-an-aardvark, is that something that someone from the TSC has to do, or should I do it?

**Comment by platinumazure** [comment]
@zertosh Assuming you know which proposal we want to focus on, you may certainly create the issue. (I haven't been following closely and am also not TSC, but I can at least assure you that the TSC does not need to drive the new issue creation.)

**Comment by dtinth** [comment]
I opened #8769.

**Comment by kaicataldo** [comment]
@dtinth Thanks for writing that up in such a detailed way!

**Comment by not-an-aardvark** [comment]
Thanks! I'm going to close this issue again in favor of #8769.
